src/dominators.rs

```diff
@@ -0,0 +1,277 @@
+                dominators: self,
+                node: self.immediate_dominator(node),
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Iterate over all of the given node's dominators (including the given
+    /// node itself).
+    ///
+    /// If the given node is not reachable from the root, then `None` is
+    /// returned.
+    pub fn dominators(&self, node: G::NodeId) -> Option<Iter<G>> {
+        if self.dominators.contains_key(&node) {
+            Some(Iter {
+                dominators: self,
+                node: Some(node),
+            })
+        } else {
+            None
+        }
+    }
+}
+
+/// Iterator for a node's dominators.
+pub struct Iter<'a, G>
+    where G: 'a + GraphBase,
+          G::NodeId: Eq + Hash
+{
+    dominators: &'a Dominators<G>,
+    node: Option<G::NodeId>,
+}
+
+impl<'a, G> Iterator for Iter<'a, G>
+    where G: 'a + GraphBase,
+          G::NodeId: Eq + Hash
+{
+    type Item = G::NodeId;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = self.node.take();
+        if let Some(next) = next {
+            self.node = self.dominators.immediate_dominator(next);
+        }
+        next
+    }
+}
+
+/// The undefined dominator sentinel, for when we have not yet discovered a
+/// node's dominator.
+const UNDEFINED: usize = ::std::usize::MAX;
+
+/// This is an implementation of the engineered ["Simple, Fast Dominance
+/// Algorithm"][0] discovered by Cooper et al.
+///
+/// This algorithm is `O(v^2)`, and therefore has slower theoretical running time
+/// than the Lenguaer-Tarjan algorithm (which is `O(e * log(v)`). However,
+/// Cooper et al found it to be faster in practice on control flow graphs of up
+/// to ~30,000 vertices.
+///
+/// [0]: http://www.cs.rice.edu/~keith/EMBED/dom.pdf
+pub fn simple_fast<G>(graph: G, root: G::NodeId) -> Dominators<G>
+    where G: IntoNeighbors + Visitable,
+          <G as GraphBase>::NodeId: Eq + Hash
+{
+    let (post_order, predecessor_sets) = simple_fast_post_order(graph, root);
+    let length = post_order.len();
+    debug_assert!(length > 0);
+    debug_assert!(post_order.last() == Some(&root));
+
+    // From here on out we use indices into `post_order` instead of actual
+    // `NodeId`s wherever possible. This greatly improves the performance of
+    // this implementation, but we have to pay a little bit of upfront cost to
+    // convert our data structures to play along first.
+
+    // Maps a node to its index into `post_order`.
+    let node_to_post_order_idx: HashMap<_, _> = post_order.iter()
+        .enumerate()
+        .map(|(idx, &node)| (node, idx))
+        .collect();
+
+    // Maps a node's `post_order` index to its set of predecessors's indices
+    // into `post_order` (as a vec).
+    let idx_to_predecessor_vec =
+        predecessor_sets_to_idx_vecs(&post_order, &node_to_post_order_idx, predecessor_sets);
+
+    let mut dominators: Vec<_> = (0..length).map(|_| UNDEFINED).collect();
```

Comment:
I *always* do this and can never remember to use that form of `vec!`, not sure why it won't stick...